src/message_bus.zig

```diff
@@ -73,60 +191,413 @@ pub const MessageBus = struct {
+
+        fn recv_header(self: *Connection) void {
+            // The fd may have been closed due to an error while sending.
+            if (self.fd == -1) return;
+            self.message_bus.io.recv(
+                *ReplicaConnection,
+                self,
+                complete_recv_header,
+                &self.completion,
+                mem.asBytes(&self.incoming_header)[self.bytes_received..],
+                os.MSG_NOSIGNAL,
+            );
+        }
+
+        fn complete_recv_header(self: *Connection, completion: *IO.Completion, result: RecvError!usize) void {
+            self.bytes_received += result catch |err| {
+                // TODO: maybe don't need to close on *every* error
+                log.err("error receiving data from {}: {}", .{ self.peer, err });
+                self.close();
+                return;
+            };
+
+            if (self.bytes_received < @sizeOf(Header)) {
+                // The header has not yet been fully received.
+                self.recv_header();
+                return;
+            }
+            assert(self.bytes_received == @sizeOf(Header));
+
+            if (!self.incoming_header.valid_checksum()) {
+                log.err("invalid checksum on header received from {}", .{self.peer});
+                self.close();
+                return;
+            }
+
+            assert(self.peer != .none);
+            if (self.peer == .unknown) {
+                if (self.incoming_header.command == .request) {
+                    self.peer = .{ .client = self.incoming_header.client };
+                } else {
+                    self.message_bus.replica_connections += 1;
+                    self.peer = .{ .replica = self.incoming_header.replica };
+                }
+            }
+
+            self.incoming_message = self.message_bus.create_message(self.incoming_header.size) catch unreachable;
+            self.incoming_message.references += 1;
+            self.incoming_message.header.* = self.incoming_header;
+            self.recv_body();
+        }
+
+        fn recv_body(self: *Connection) void {
+            // The fd may have been closed due to an error while sending.
+            if (self.fd == -1) return;
+            self.message_bus.io.recv(
+                *ReplicaConnection,
+                self,
+                complete_recv_body,
+                &self.completion,
+                self.incoming_message.buffer[self.bytes_received..],
+                os.MSG_NOSIGNAL,
+            );
+        }
+
+        fn complete_recv_body(self: *Connection, completion: *IO.Completion, result: RecvError!usize) void {
+            self.bytes_received += result catch |err| {
+                // TODO: maybe don't need to close on *every* error
+                log.err("error receiving data from: {}", .{ peer, err });
+                self.close();
+                return;
+            };
+
+            if (self.bytes_received < self.incoming_message.buffer.len) {
```

Comment:
We should compare against `message.header.size` to be future-proof since with static allocation our buffer sizes will mostly be slightly bigger or much bigger than the actual message.